src/ray/stats/stats.h

```diff
@@ -38,38 +39,72 @@ namespace stats {
+/// - stats::Init should be called only once per PROCESS. Redundant calls will be just
+/// ignored.
+/// - If you want to reinitialize, you should call stats::Shutdown().
+/// - It is thread-safe.
+/// We recommend you to use this only once inside a main script and add Shutdown() method
+/// to any signal handler. \param global_tags Tags that will be appended to all metrics in
+/// this process. \param metrics_agent_port The port to export metrics at each node.
+/// \param exporter_to_use The exporter client you will use for this process' metrics.
+static void Init(const TagsType &global_tags, const int metrics_agent_port,
+                 std::shared_ptr<MetricExporterClient> exporter_to_use = nullptr,
+                 int64_t metrics_report_batch_size =
+                     RayConfig::instance().metrics_report_batch_size()) {
+  absl::MutexLock lock(&stats_mutex);
+  if (StatsConfig::instance().IsInitialized()) {
+    RAY_CHECK(metrics_io_service_pool != nullptr);
+    RAY_CHECK(exporter != nullptr);
+    return;
+  }
+
+  RAY_CHECK(metrics_io_service_pool == nullptr);
+  RAY_CHECK(exporter == nullptr);
+  bool disable_stats = !RayConfig::instance().enable_metrics_collection();
   StatsConfig::instance().SetIsDisableStats(disable_stats);
   if (disable_stats) {
     RAY_LOG(INFO) << "Disabled stats.";
     return;
   }
 
-  // Force to have a singleton exporter.
-  static std::shared_ptr<MetricExporterClient> exporter;
-  // Default exporter is metrics agent exporter.
+  metrics_io_service_pool = std::make_shared<IOServicePool>(1);
+  metrics_io_service_pool->Run();
+  boost::asio::io_service *metrics_io_service = metrics_io_service_pool->Get();
+  RAY_CHECK(metrics_io_service != nullptr);
+
+  // Default exporter is a metrics agent exporter.
   if (exporter_to_use == nullptr) {
     std::shared_ptr<MetricExporterClient> stdout_exporter(new StdoutExporterClient());
     exporter.reset(new MetricsAgentExporter(stdout_exporter, metrics_agent_port,
-                                            io_service, "127.0.0.1"));
+                                            (*metrics_io_service), "127.0.0.1"));
   } else {
     exporter = exporter_to_use;
   }
 
-  // TODO(sang): Currently, we don't do any cleanup. This can lead us to lose last 10
-  // seconds data before we exit the main script.
   MetricExporter::Register(exporter, metrics_report_batch_size);
   opencensus::stats::StatsExporter::SetInterval(
       StatsConfig::instance().GetReportInterval());
   opencensus::stats::DeltaProducer::Get()->SetHarvestInterval(
       StatsConfig::instance().GetHarvestInterval());
   StatsConfig::instance().SetGlobalTags(global_tags);
+  StatsConfig::instance().SetIsInitialized(true);
+}
+
+/// Shutdown the initialized stats library.
+/// This cleans up various threads and metadata for stats library.
+static void Shutdown() {
+  absl::MutexLock lock(&stats_mutex);
+  opencensus::stats::StatsExporter::Shutdown();
+  metrics_io_service_pool->Stop();
```

Comment:
Metric io service should be stopped before stats exporter shutdown because async operations will be taken and access bad memory address.

Reply:
Good catch! let me fix it.